Firestore/core/src/firebase/firestore/core/sync_engine.mm

```diff
@@ -0,0 +1,523 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/core/sync_engine.h"
+
+#include "Firestore/core/include/firebase/firestore/firestore_errors.h"
+#include "Firestore/core/src/firebase/firestore/core/transaction.h"
+#include "Firestore/core/src/firebase/firestore/core/transaction_runner.h"
+#include "Firestore/core/src/firebase/firestore/local/query_data.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/document_map.h"
+#include "Firestore/core/src/firebase/firestore/model/document_set.h"
+#include "Firestore/core/src/firebase/firestore/model/no_document.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+
+namespace firebase {
+namespace firestore {
+namespace core {
+
+namespace {
+
+using firestore::Error;
+using model::ListenSequenceNumber;
+using util::Status;
+
+// Limbo documents don't use persistence, and are eagerly GC'd. So, listens for
+// them don't need real sequence numbers.
+const ListenSequenceNumber kIrrelevantSequenceNumber = -1;
+
+bool ErrorIsInteresting(const Status& error) {
+  bool missing_index =
+      (error.code() == Error::FailedPrecondition &&
+       error.error_message().find("requires an index") != std::string::npos);
+  bool no_permission = (error.code() == Error::PermissionDenied);
+  return missing_index || no_permission;
+}
+
+}  // namespace
+
+using auth::User;
+using local::LocalViewChanges;
```

Comment:
Final nit: all `using` declarations should be in the unnamed namespace (move them up, to where `using firestore::Error` is). See the TotW I linked in the other reply.

Reply:
Done.